### ISSUE TITLE
Conflicting MQTT-Versions lead to leading 0-byte payload in case of using persistence

### ISSUE THREAD

**Issue by ghost** [eclipse/paho.mqtt.cpp#261]
I figured out an inconsistency in the MQTT-versions stored in the client and in the opts which leads to an error when using persistence. The error is that the payload of a persisted message is send with a leading 0-byte.

Which implementation of the `mqtt::iclient_persistence` interface is used has no influence on the issue. The important thing is, that the error only occurs if the `get()`-function of the persistence is used. Blocking the network while publishing messages and then unblocking is not sufficient to reproduce the error. The `get()`-function is only called if the client is restarted (stop/start process) 

The issue is a bit tricky to explain... but I'll try my best:

### The inconsistency is as follows:

- async-client (async_client.cpp) constructors - async_client::async_client:
Through the initialization list of the constructors the member variable `mqttVersion_` is set to `MQTTVERSION_DEFAULT` (0). But in the constructor implementation in the first line (`MQTTAsync_createOptions opts MQTTAsync_createOptions_initializer5;`) the `opts_.MQTTVersion` is set to `MQTTVERSION_5` (5). This opts are used to create the underlying C-client (`int rc = MQTTAsync_createWithOptions(... &opts)`)
Thus:
- The underlying C-client is initialized with `MQTTVERSION_5` (hard-coded)
- The C++-client (`mqttVersion_`) is initialized with `MQTTVERSION_DEFAULT`

- async-client (async_client.cpp) - async_client::connect
The member variable `mqttVersion_` is set to `opts.opts_.MQTTVersion` where the opts are provided as a parameter. In my case (as I cannot use MQTT v5) this again is `MQTTVERSION_DEFAULT`.

### Now the issue happens in case that a stored message is restored. 
1. Block the connection to the MQTT Broker
2. Send a message
   - the message is stored by the persistance implementation (e.g. in a file)
3. Stop the process where the client is running in
4. Start the process again
   --> Restarting the process forces, that the message is read by the `get()` function of the persistence implementation (interface: `mqtt::iclient_persistence`, function: `mqtt::string_view get(const std::string& key)` )

In that case the following function of the underlying C-client is executed: 
`int MQTTPersistence_restore(Clients *c)`.
This function has the following code:
```
publish->MQTTVersion = c->MQTTVersion;
msg = MQTTProtocol_createMessage(publish, &msg, publish->header.bits.qos, publish->header.bits.retain);
```

--> `publish->MQTTVersion` is set to `c->MQTTVersion`, which is the clients MQTT-version, which is "hard-coded" to `MQTTVERSION_5`

Now, when publishing this message the error occurs in MQTTPacket.c - function `MQTTPacket_send_publish`
source line:
```int buflen = ((qos > 0) ? 2 : 0) + ((pack->MQTTVersion >= 5) ? MQTTProperties_len(&pack->properties) : 0);```
As `qos` is 1 and the `pack->MQTTVersion` is equals to 5 and the `MQTTProperties_len(&pack->properties)` is 1 the resulting buflen is 3 !!! But it should be 2!

As buflen is 3, bufs[2] is 3-chars in size and thus, bufs[2][2] is a 0-byte.
As in the remaining processing of the message the buflen is 2, this 0-byte is interpreted as the first char in the bufs[3], which is the payload.

**Comment by AndreasSteck** [comment]
I am using:
paho.mqtt.c: v1.3.1
paho.mqtt.cpp: v1.1

**Comment by AndreasSteck** [comment]
This is the persistence implementation I used with some "debugging" cout's added:
```
class sample_file_persistence : virtual public mqtt::iclient_persistence
{
private:

    const std::string PERSIST_FILENAME = "persistMqtt";

    std::fstream persistFileStream;

	// Use an STL map to store shared persistence pointers
	// against string keys.
	std::map<std::string, std::string> store_;

    void storeMapToFile() 
    {
        std::cout << "Store map to file (size = " << store_.size() << ")" << std::endl;

        std::ofstream persistFileStream(PERSIST_FILENAME, std::ofstream::trunc | std::ofstream::binary );
        
        boost::archive::binary_oarchive oarch(persistFileStream);
        oarch << store_;
        persistFileStream << std::flush;

        persistFileStream.close();
    }

    void loadMapFromFile() 
    {     
        std::ifstream persistFileStream(PERSIST_FILENAME, std::ifstream::binary );
        if(persistFileStream)
        {
            boost::archive::binary_iarchive iarch(persistFileStream);
            iarch >> store_;
            std::cout << "Load map from file (size = " << store_.size() << ")" << std::endl;
        }
        else
        {
            std::cout << "Load map -- no file provided" << std::endl;
        }
    }

public:
	// Open the persistent store
	void open(const std::string& clientId, const std::string& serverURI) override 
    {
        loadMapFromFile();
		std::cout << "[Opening persistence store for '" << clientId
			<< "' at '" << serverURI << "']" << std::endl;
	}

	// Close the persistent store
	void close() override 
    {
        storeMapToFile();
		std::cout << "[Closing persistence store.]" << std::endl;
	}

	// Clears persistence
	void clear() override 
    {
		std::cout << "[Clearing persistence store.] --> do nothing" << std::endl;
	}

	// Returns whether or not data is persisted using the specified key.
	bool contains_key(const std::string &key) override 
    {
        return store_.find(key) != store_.end();
	}

	// Returns the keys in this persistent data store.
	const mqtt::string_collection& keys() const override 
    {

        static mqtt::string_collection ks;
        ks.clear();

        for (const auto& k : store_) 
        {
            ks.push_back(k.first);
        }

        return ks;
	}

	// Puts the specified data into the persistent store.
	void put(const std::string& key, const std::vector<mqtt::string_view>& bufs) override 
    {
        std::string str;
        std::cout << "persistence - put - chars: ";
        for (const auto& b : bufs) 
        {
            for(int i=0; i<b.size(); i++)
            {
                str += b[i];
                printf(" %d", b[i]);
            }
        }
        std::cout << std::endl;

        std::cout << "persistence - put [key][value]: [" << key << "][" << str << "]" << std::endl;
        store_[key] = std::move(str);

        storeMapToFile();
	}

	// Gets the specified data out of the persistent store.
	mqtt::string_view get(const std::string& key) const override 
    {
        auto p = store_.find(key);
        if (p == store_.end()) 
        {
            throw mqtt::persistence_exception();
        }
        std::cout << "persistence - get [key][value]: [" << key << "][" << p->second << "]" << std::endl;
        std::cout << "persistence - get - chars: ";
        for(const char &c: p->second)
        {
            printf(" %d", c);
        }
        std::cout << std::endl;

        return mqtt::string_view(p->second);
	}

	// Remove the data for the specified key.
	void remove(const std::string &key) override 
    {
        auto p = store_.find(key);
        if (p == store_.end()) 
        {
            throw mqtt::persistence_exception();
        }
        store_.erase(p);

        storeMapToFile();
	}
};
```

**Comment by AndreasSteck** [comment]
And here the outputs:
1. Start the client
2. Block the connection to the MQTT Broker
3. Publish payload ABCD

persistence - put - chars:  3 0 0 0 2 0 0 0 47 100 101 118 105 99 101 115 47 82 48 48 48 49 47 101 118 101 110 116 115 47 101 120 97 109 112 108 101 45 116 111 112 105 99 0 4 0 0 0 65 66 67 68 1 0 0 0 0 0 0 0
persistence - put [key][value]: [c-2][/devices/R0001/events/example-topicABCD]

persistence - put - chars:  50 43 0 35 47 100 101 118 105 99 101 115 47 82 48 48 48 49 47 101 118 101 110 116 115 47 101 120 97 109 112 108 101 45 116 111 112 105 99 **`0 2`** 65 66 67 68
persistence - put [key][value]: [s-2][2+#/devices/R0001/events/example-topicABCD]

4. Stop the process
5. Unblock the connection to the Broker
6. Start the process

persistence - get [key][value]: [s-2][2+#/devices/R0001/events/example-topicABCD]
persistence - get - chars:  50 43 0 35 47 100 101 118 105 99 101 115 47 82 48 48 48 49 47 101 118 101 110 116 115 47 101 120 97 109 112 108 101 45 116 111 112 105 99 **`0 2`** 65 66 67 68
--> get reads is correct: **`0 2`** 65 66 67 68

persistence - put - chars:  58 44 0 35 47 100 101 118 105 99 101 115 47 82 48 48 48 49 47 101 118 101 110 116 115 47 101 120 97 109 112 108 101 45 116 111 112 105 99 **`0 2 0`** 65 66 67 68
persistence - put [key][value]: [s5-2][:,#/devices/R0001/events/example-topicABCD]
--> Here is the error: **`0 2 0`** 65 66 67 68
--> The payload is send as �ABCD (0 65 66 67 68)

**Comment by fpagliughi** [comment]
This was fixed in v1.2, and  is being further updated for v1.3. The recommendation is to create the options for the desired version and not change it.
